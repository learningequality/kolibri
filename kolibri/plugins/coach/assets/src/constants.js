const PageNames = {
  CLASS_LIST: 'CLASS_LIST',
  RECENT_CHANNELS: 'RECENT_CHANNELS',
  RECENT_ITEMS_FOR_CHANNEL: 'RECENT_ITEMS_FOR_CHANNEL',
  RECENT_LEARNERS_FOR_ITEM: 'RECENT_LEARNERS_FOR_ITEM',
  RECENT_LEARNER_ITEM_DETAILS: 'RECENT_LEARNER_ITEM_DETAILS',
  RECENT_LEARNER_ITEM_DETAILS_ROOT: 'RECENT_LEARNER_ITEM_DETAILS_ROOT',
  TOPIC_CHANNELS: 'TOPIC_CHANNELS',
  TOPIC_CHANNEL_ROOT: 'TOPIC_CHANNEL_ROOT',
  TOPIC_ITEM_LIST: 'TOPIC_ITEM_LIST',
  TOPIC_LEARNERS_FOR_ITEM: 'TOPIC_LEARNERS_FOR_ITEM',
  TOPIC_LEARNER_ITEM_DETAILS: 'TOPIC_LEARNER_ITEM_DETAILS',
  TOPIC_LEARNER_ITEM_DETAILS_ROOT: 'TOPIC_LEARNER_ITEM_DETAILS_ROOT',
  LEARNER_LIST: 'LEARNER_LIST',
  LEARNER_CHANNELS: 'LEARNER_CHANNELS',
  LEARNER_CHANNEL_ROOT: 'LEARNER_CHANNEL_ROOT',
  LEARNER_ITEM_LIST: 'LEARNER_ITEM_LIST',
  LEARNER_ITEM_DETAILS: 'LEARNER_ITEM_DETAILS',
  LEARNER_ITEM_DETAILS_ROOT: 'LEARNER_ITEM_DETAILS_ROOT',
  EXAMS: 'EXAMS',
  CREATE_EXAM: 'CREATE_EXAM',
  EXAM_REPORT: 'EXAM_REPORT',
  EXAM_REPORT_DETAIL: 'EXAM_REPORT_DETAIL',
  EXAM_REPORT_DETAIL_ROOT: 'EXAM_REPORT_DETAIL_ROOT',
  LEARNERS: 'LEARNERS',
  GROUPS: 'GROUPS',
};

const PageTitles = {
  EXAMS: 'Exams',
  CREATE_EXAM: 'Create new exam',
};

const RecentReports = [
  PageNames.RECENT_CHANNELS,
  PageNames.RECENT_ITEMS_FOR_CHANNEL,
  PageNames.RECENT_LEARNERS_FOR_ITEM,
  PageNames.RECENT_LEARNER_ITEM_DETAILS,
];

const TopicReports = [
  PageNames.TOPIC_CHANNELS,
  PageNames.TOPIC_CHANNEL_ROOT,
  PageNames.TOPIC_ITEM_LIST,
  PageNames.TOPIC_LEARNERS_FOR_ITEM,
  PageNames.TOPIC_LEARNER_ITEM_DETAILS,
];

const LearnerReports = [
  PageNames.LEARNER_LIST,
  PageNames.LEARNER_CHANNELS,
  PageNames.LEARNER_CHANNEL_ROOT,
  PageNames.LEARNER_ITEM_LIST,
  PageNames.LEARNER_ITEM_DETAILS,
];

const ExamPages = [
  PageNames.EXAMS,
  PageNames.CREATE_EXAM,
  PageNames.EXAM_REPORT,
  PageNames.EXAM_REPORT_DETAIL,
  PageNames.LEARNERS,
];

const GroupModals = {
  CREATE_GROUP: 'CREATE_GROUP',
  RENAME_GROUP: 'RENAME_GROUP',
  DELETE_GROUP: 'DELETE_GROUP',
  MOVE_LEARNERS: 'MOVE_LEARNERS',
};

export {
  PageNames,
  PageTitles,
  RecentReports,
  TopicReports,
  ExamPages,
  LearnerReports,
  GroupModals,
};
