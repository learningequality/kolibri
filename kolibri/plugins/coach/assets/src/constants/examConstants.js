export const Modals = {
  PREVIEW_EXAM: 'PREVIEW_EXAM',
};
