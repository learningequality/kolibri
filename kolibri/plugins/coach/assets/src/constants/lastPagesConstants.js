export const LastPages = {
  HOME_PAGE: 'homepage',
  HOME_ACTIVITY: 'homeactivity',
  GROUP_ACTIVITY: 'groupactivity',
  LEARNER_ACTIVITY: 'learneractivity',
  EXERCISE_LEARNER_LIST: 'exerciselearnerlist',
  EXERCISE_LEARNER_LIST_BY_GROUPS: 'exerciselearnerlistbygroups',
  RESOURCE_LEARNER_LIST: 'resourcelearnerlist',
  RESOURCE_LEARNER_LIST_BY_GROUPS: 'resourcelearnerlistbygroups',
};
