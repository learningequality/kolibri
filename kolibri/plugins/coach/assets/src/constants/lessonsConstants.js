export const LessonsPageNames = {
  PLAN_LESSONS_ROOT: 'PLAN_LESSONS_ROOT',
  LESSON_CREATION_ROOT: 'LESSON_CREATION_ROOT',
  SUMMARY: 'SUMMARY',
  RESOURCE_USER_SUMMARY: 'RESOURCE_USER_SUMMARY',
  RESOURCE_USER_REPORT: 'RESOURCE_USER_REPORT',
  RESOURCE_USER_REPORT_ROOT: 'RESOURCE_USER_REPORT_ROOT',
  RESOURCE_CONTENT_PREVIEW: 'RESOURCE_CONTENT_PREVIEW', // exclusively a route name
  SELECTION_ROOT: 'SELECTION_ROOT',
  SELECTION: 'SELECTION',
  SELECTION_SEARCH: 'SELECTION_SEARCH',
  SELECTION_CONTENT_PREVIEW: 'SELECTION_CONTENT_PREVIEW', // exclusively a route name
  CONTENT_PREVIEW: 'CONTENT_PREVIEW',
};

export const CollectionTypes = {
  LEARNERGROUP: 'learnergroup',
  CLASSROOM: 'classroom',
};
