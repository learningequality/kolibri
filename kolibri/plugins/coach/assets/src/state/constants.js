const PageNames = {
  COACH_CLASS_LIST_PAGE: 'COACH_CLASS_LIST_PAGE',
  COACH_RECENT_PAGE: 'COACH_RECENT_PAGE',
  COACH_TOPICS_PAGE: 'COACH_TOPICS_PAGE',
  COACH_EXAMS_PAGE: 'COACH_EXAMS_PAGE',
  COACH_LEARNERS_PAGE: 'COACH_LEARNERS_PAGE',
  COACH_GROUPS_PAGE: 'COACH_GROUPS_PAGE',
  COACH_CONTENT_LEARNERS_PAGE: 'COACH_CONTENT_LEARNERS_PAGE',
};

const ContentScopes = {
  ROOT: 'root',
  TOPIC: 'topic',
  CONTENT: 'content',
};

const UserScopes = {
  FACILITY: 'facility',
  CLASSROOM: 'classroom',
  LEARNERGROUP: 'learnergroup',
  USER: 'user',
};

const AllOrRecent = {
  ALL: 'all',
  RECENT: 'recent',
};

const ViewBy = {
  LEARNERS: 'user_view',
  CONTENT: 'content_view',
};

const TableColumns = {
  NAME: 'name',
  EXERCISE: 'exercise_progress',
  CONTENT: 'content_progress',
  DATE: 'date',
};

const SortOrders = {
  ASCENDING: 'asc',
  DESCENDING: 'desc',
  NONE: 'none',
};

module.exports = {
  PageNames,
  ContentScopes,
  UserScopes,
  AllOrRecent,
  ViewBy,
  TableColumns,
  SortOrders,
};
