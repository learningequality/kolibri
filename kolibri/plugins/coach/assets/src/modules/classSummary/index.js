import get from 'lodash/get';
import set from 'lodash/set';
import flatten from 'lodash/flatten';
import find from 'lodash/find';
import Vue from 'kolibri.lib.vue';
import ClassSummaryResource from '../../apiResources/classSummary';
import dataHelpers from './dataHelpers';
import { STATUSES } from './constants';
import { updateWithNotifications } from './actions';

function defaultState() {
  return {
    id: null,
    name: '',
    /*
     * coachMap := {
     *   [id]: { id, name, username }
     * }
     */
    coachMap: {},
    /*
     * learnerMap := {
     *   [id]: { id, name, username }
     * }
     */
    learnerMap: {},
    /*
     * groupMap := {
     *   [id]: { id, name, member_ids: [id, ...] }
     * }
     */
    groupMap: {},
    /*
     * examMap := {
     *   [id]: {
     *     id,
     *     active,
     *     title,
     *     question_sources: [{exercise_id, question_id}, ...],
     *     groups: [id, ...],
     *     data_model_version,
     *     question_count,
     *   }
     * }
     */
    adHocGroupsMap: {},
    /*
     * adHocLearners := {
     *  [collection_id]: {
     *    id,
     *    member_ids: [user_id, ...]
     *  },
     * }
     */
    examMap: {},
    /*
     * examLearnerStatusMap := {
     *   [exam_id]: {
     *     [learner_id]: { exam_id, learner_id, status, last_activity, num_correct, score }
     *   }
     * }
     */
    examLearnerStatusMap: {},
    /*
     * contentMap := {
     *   [content_id]: { content_id, node_id, kind, title }
     * }
     */
    contentMap: {},
    /*
     * contentNodeMap := {
     *   [node_id]: { content_id, node_id, kind, title }
     * }
     */
    contentNodeMap: {},
    /*
     * contentLearnerStatusMap := {
     *   [content_id]: {
     *     [learner_id]: { content_id, learner_id, status, last_activity }
     *   }
     * }
     */
    contentLearnerStatusMap: {},
    /*
     * lessonMap := {
     *   [id]: { id, active, title, node_ids: [id, ...], groups: [id, ...] }
     * }
     */
    lessonMap: {},
  };
}

// return a map of keys to items
export function _itemMap(items, key) {
  const itemMap = {};
  items.forEach(item => {
    itemMap[item[key]] = item;
  });
  return itemMap;
}

// return a map of keys to maps of learner ids to statuses
export function _statusMap(statuses, key) {
  const statusMap = {};
  statuses.forEach(status => {
    if (!statusMap[status[key]]) {
      statusMap[status[key]] = {};
    }
    statusMap[status[key]][status.learner_id] = status;
  });
  return statusMap;
}

function _lessonStatusForLearner(state, lessonId, learnerId) {
  const lesson = state.lessonMap[lessonId];
  // Add trap door for lesson status, if a lesson has no resources
  // make it impossible for a learner to be registered as anything
  // but not started.
  if (lesson.node_ids.length === 0) {
    return STATUSES.notStarted;
  }
  const statuses = lesson.node_ids.map(node_id => {
    if (!state.contentNodeMap[node_id]) {
      return { status: STATUSES.notStarted };
    }
    const content_id = state.contentNodeMap[node_id].content_id;
    return {
      status: get(
        state.contentLearnerStatusMap,
        [content_id, learnerId, 'status'],
        STATUSES.notStarted
      ),
    };
  });

  const tally = {
    [STATUSES.started]: 0,
    [STATUSES.notStarted]: 0,
    [STATUSES.completed]: 0,
    [STATUSES.helpNeeded]: 0,
  };
  statuses.forEach(status => {
    tally[status.status] += 1;
  });
  if (tally[STATUSES.helpNeeded]) {
    return STATUSES.helpNeeded;
  }
  if (tally[STATUSES.completed] === statuses.length) {
    return STATUSES.completed;
  }
  if (tally[STATUSES.notStarted] === statuses.length) {
    return STATUSES.notStarted;
  }
  return STATUSES.started;
}

// convert quiz scores to percentages from integer counts of correct answers
function _score(numCorrect, numQuestions) {
  if (numCorrect === null || numCorrect === undefined) {
    return null;
  }
  return (1.0 * numCorrect) / numQuestions;
}

export default {
  namespaced: true,
  state: defaultState(),
  getters: {
    ...dataHelpers,
    /*
     * coaches := [
     *   { id, name, username }, ...
     * ]
     */
    coaches(state) {
      return Object.values(state.coachMap);
    },
    /*
     * learners := [
     *   { id, name, username }, ...
     * ]
     */
    learners(state) {
      return Object.values(state.learnerMap);
    },
    /*
     * adHocGroups := [
     *   { id, member_ids: [id, ...] }, ...
     * ]
     ]
     */
    adHocGroups(state) {
      return Object.values(state.adHocGroupsMap);
    },
    /*
     * groups := [
     *   { id, name, member_ids: [id, ...] }, ...
     * ]
     */
    groups(state) {
      return Object.values(state.groupMap);
    },
    /*
     * exams := [
     *   {
     *     id,
     *     active,
     *     title,
     *     question_sources: [{exercise_id, question_id}, ...],
     *     groups: [id, ...],
     *   },
     *   ...
     * ]
     */
    exams(state) {
      return Object.values(state.examMap);
    },
    /*
     * examStatuses := [
     *   { exam_id, learner_id, status, last_activity }, ...
     * ]
     */
    examStatuses(state) {
      return flatten(
        Object.values(state.examLearnerStatusMap).map(learnerMap => Object.values(learnerMap))
      );
    },
    /*
     * content := [
     *   { content_id, node_id, kind, title }, ...
     * ]
     */
    content(state) {
      return Object.values(state.contentMap);
    },
    /*
     * contentStatuses := [
     *   { content_id, learner_id, status, last_activity }, ...
     * ]
     */
    contentStatuses(state) {
      return flatten(
        Object.values(state.contentLearnerStatusMap).map(learnerMap => Object.values(learnerMap))
      );
    },
    /*
     * lessons := [
     *   { id, active, title, node_ids: [id, ...], groups: [id, ...] }, ...
     * ]
     */
    lessons(state) {
      return Object.values(state.lessonMap);
    },
    /*
     * lessonStatuses := [
     *   { lesson_id, learner_id, status, last_activity }, ...
     * ]
     */
    lessonStatuses(state, getters) {
      return flatten(
        Object.values(getters.lessonLearnerStatusMap).map(learnerMap => Object.values(learnerMap))
      );
    },
    /*
     * lessonLearnerStatusMap := {
     *   [lesson_id]: {
     *     [learner_id]: { lesson_id, learner_id, status, last_activity }
     *   }
     * }
     */
    lessonLearnerStatusMap(state) {
      const map = {};
      Object.values(state.lessonMap).forEach(lesson => {
        map[lesson.id] = {};
        Object.values(state.learnerMap).forEach(learner => {
          let last = null;
          // for all content items this learner has interacted with
          // determine the latest interaction activity
          Object.values(lesson.node_ids).forEach(node_id => {
            const content_id = get(state.contentNodeMap, [node_id, 'content_id'], null);
            const last_activity = get(
              state.contentLearnerStatusMap,
              [content_id, learner.id, 'last_activity'],
              null
            );
            if (last_activity > last) {
              last = last_activity;
            }
          });
          map[lesson.id][learner.id] = {
            lesson_id: lesson.id,
            learner_id: learner.id,
            status: _lessonStatusForLearner(state, lesson.id, learner.id),
            last_activity: last,
          };
        });
      });
      return map;
    },
    quizTitleUnavailable(state, getters) {
      const normalize = title => title.trim().toUpperCase();
      return function finder({ title, excludeId }) {
        return find(
          getters.exams,
          exam => exam.id !== excludeId && normalize(exam.title) === normalize(title)
        );
      };
    },
    lessonTitleUnavailable(state, getters) {
      const normalize = title => title.trim().toUpperCase();
      return function finder({ title, excludeId }) {
        return find(
          getters.lessons,
          lesson => lesson.id !== excludeId && normalize(lesson.title) === normalize(title)
        );
      };
    },
    // Adapter used in 'coachNotifications' module. Make sure this getter is updated
    // whenever this module's state changes.
    notificationModuleData(state) {
      return {
        adHocGroupsMap: state.adHocGroupsMap,
        learners: state.learnerMap,
        learnerGroups: state.groupMap,
        lessons: state.lessonMap,
        exams: state.examMap,
        classId: state.id,
        className: state.name,
        contentNodes: state.contentNodeMap,
      };
    },
  },
  mutations: {
    SET_STATE(state, summary) {
      const examMap = _itemMap(summary.exams, 'id');
      const lessonMap = _itemMap(summary.lessons, 'id');
      Object.values(examMap).forEach(exam => {
        // convert dates
        exam.date_created = new Date(exam.date_created);
      });
      Object.values(lessonMap).forEach(lesson => {
        // convert dates
        lesson.date_created = new Date(lesson.date_created);
      });
      summary.exam_learner_status.forEach(status => {
        // convert dates
        status.last_activity = status.last_activity ? new Date(status.last_activity) : null;
        status.score = _score(status.num_correct, examMap[status.exam_id].question_count);
      });
      summary.content_learner_status.forEach(status => {
        // convert dates
        status.last_activity = status.last_activity ? new Date(status.last_activity) : null;
      });
      const patchedSummaryContent = summary.content.map(item => {
        const obj = Object.assign({}, item);
        obj['kind'] = obj['kind'] == 'h5p' ? 'html5' : obj['kind'];
        return obj;
      });
      Object.assign(state, {
        id: summary.id,
        facility_id: summary.facility_id,
        name: summary.name,
        coachMap: _itemMap(summary.coaches, 'id'),
        learnerMap: _itemMap(summary.learners, 'id'),
        groupMap: _itemMap(summary.groups, 'id'),
        adHocGroupsMap: _itemMap(summary.adhoclearners, 'id'),
        examMap,
        examLearnerStatusMap: _statusMap(summary.exam_learner_status, 'exam_id'),
        contentMap: _itemMap(patchedSummaryContent, 'content_id'),
        contentNodeMap: _itemMap(patchedSummaryContent, 'node_id'),
        contentLearnerStatusMap: _statusMap(summary.content_learner_status, 'content_id'),
        lessonMap,
      });
    },
    CREATE_ITEM(state, { map, id, object }) {
      state[map] = {
        ...state[map],
        [id]: object,
      };
    },
    UPDATE_ITEM(state, { map, id, object }) {
      Object.assign(state[map][id], object);
    },
    DELETE_ITEM(state, { map, id }) {
      Vue.delete(state[map], id);
    },
    APPLY_NOTIFICATION_UPDATES(state, updates) {
      const { contentLearnerStatusMap, examLearnerStatusMap } = state;
      const { contentLearnerStatusMapUpdates, examLearnerStatusMapUpdates } = updates;

      contentLearnerStatusMapUpdates.forEach(update => {
        const path = [update.content_id, update.learner_id];
        const currentStatus = get(contentLearnerStatusMap, path);
        if (currentStatus) {
          Object.assign(currentStatus, update);
        } else {
          set(state.contentLearnerStatusMap, path, {
            ...update,
            time_spent: 0,
          });
        }
      });

      examLearnerStatusMapUpdates.forEach(update => {
        const path = [update.exam_id, update.learner_id];
        const scoredUpdate = {
          ...update,
          score: _score(update.num_correct, state.examMap[update.exam_id].question_count),
        };
        const currentStatus = get(examLearnerStatusMap, path);
        if (currentStatus) {
          Object.assign(currentStatus, scoredUpdate);
        } else {
          set(state.examLearnerStatusMap, path, scoredUpdate);
        }
      });

      state.examLearnerStatusMap = { ...state.examLearnerStatusMap };
      state.contentLearnerStatusMap = { ...state.contentLearnerStatusMap };
    },
  },
  actions: {
    updateWithNotifications,
    loadClassSummary(store, classId) {
      return ClassSummaryResource.fetchModel({ id: classId, force: true }).then(summary => {
        store.commit('SET_STATE', summary);
      });
    },
    refreshClassSummary(store) {
      return store.dispatch('loadClassSummary', store.state.id);
    },
  },
};
