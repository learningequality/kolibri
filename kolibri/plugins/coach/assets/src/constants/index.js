import { LessonsPageNames } from './lessonsConstants';

export const PageNames = {
  REPORTS_PAGE: 'REPORTS_PAGE',
  PLAN_PAGE: 'PLAN_PAGE',
  NEW_COACH_PAGES: 'NEW_COACH_PAGES',
  EXAMS: 'EXAMS',
  EXAM_CREATION_ROOT: 'EXAM_CREATION_ROOT',
  EXAM_CREATION_CHANNEL_QUIZ: 'EXAM_CREATION_CHANNEL_QUIZ',
  EXAM_CREATION_SELECT_CHANNEL_QUIZ_TOPIC: 'EXAM_CREATION_SELECT_CHANNEL_QUIZ_TOPIC',
  EXAM_CREATION_CHANNEL_QUIZ_PREVIEW: 'EXAM_CREATION_CHANNEL_QUIZ_PREVIEW',
  EXAM_CREATION_TOPIC: 'EXAM_CREATION_TOPIC',
  EXAM_CREATION_BOOKMARKS: 'EXAM_CREATION_BOOKMARKS',
  EXAM_CREATION_BOOKMARKS_MAIN: 'EXAM_CREATION_BOOKMARKS_MAIN',
  EXAM_CREATION_PREVIEW: 'EXAM_CREATION_PREVIEW',
  EXAM_CREATION_SEARCH: 'EXAM_CREATION_SEARCH',
  EXAM_CREATION_QUESTION_SELECTION: 'EXAM_CREATION_QUESTION_SELECTION',
  EXAM_PREVIEW: 'EXAM_PREVIEW',
  EXAM_REPORT: 'EXAM_REPORT',
  EXAM_REPORT_DETAIL: 'EXAM_REPORT_DETAIL',
  EXAM_REPORT_DETAIL_ROOT: 'EXAM_REPORT_DETAIL_ROOT',
  REPORTS_LESSON_EXERCISE_LEARNER_PAGE_ROOT: 'REPORTS_LESSON_EXERCISE_LEARNER_PAGE_ROOT',
  REPORTS_GROUP_REPORT_LESSON_EXERCISE_LEARNER_PAGE_ROOT:
    'REPORTS_GROUP_REPORT_LESSON_EXERCISE_LEARNER_PAGE_ROOT',
  REPORTS_LEARNER_ACTIVITY_EXERCISE_PAGE_ROOT: 'REPORTS_LEARNER_ACTIVITY_EXERCISE_PAGE_ROOT',
  REPORTS_LEARNER_REPORT_LESSON_EXERCISE_PAGE_ROOT:
    'REPORTS_LEARNER_REPORT_LESSON_EXERCISE_PAGE_ROOT',
  REPORTS_LESSON_LEARNER_EXERCISE_PAGE_ROOT: 'REPORTS_LESSON_LEARNER_EXERCISE_PAGE_ROOT',
  REPORTS_GROUP_LEARNER_REPORT_QUIZ_PAGE_ROOT: 'REPORTS_GROUP_LEARNER_REPORT_QUIZ_PAGE_ROOT',
  REPORTS_GROUP_REPORT_QUIZ_LEARNER_PAGE_ROOT: 'REPORTS_GROUP_REPORT_QUIZ_LEARNER_PAGE_ROOT',
  REPORTS_LEARNER_REPORT_QUIZ_PAGE_ROOT: 'REPORTS_LEARNER_REPORT_QUIZ_PAGE_ROOT',
  REPORTS_QUIZ_LEARNER_PAGE_ROOT: 'REPORTS_QUIZ_LEARNER_PAGE_ROOT',
  REPORTS_LESSON_EXERCISE_QUESTION_PAGE_ROOT: 'REPORTS_LESSON_EXERCISE_QUESTION_PAGE_ROOT',
  REPORTS_LESSON_PRACTICE_QUIZ_QUESTION_PAGE_ROOT:
    'REPORTS_LESSON_PRACTICE_QUIZ_QUESTION_PAGE_ROOT',
  REPORTS_GROUP_REPORT_LESSON_EXERCISE_QUESTION_PAGE_ROOT:
    'REPORTS_GROUP_REPORT_LESSON_EXERCISE_QUESTION_PAGE_ROOT',
  REPORTS_GROUP_REPORT_LESSON_PRACTICE_QUIZ_QUESTION_PAGE_ROOT:
    'REPORTS_GROUP_REPORT_LESSON_PRACTICE_QUIZ_QUESTION_PAGE_ROOT',
  REPORTS_GROUP_REPORT_LESSON_PRACTICE_QUIZ_LEARNER_PAGE_ROOT:
    'REPORTS_GROUP_REPORT_LESSON_PRACTICE_QUIZ_LEARNER_PAGE_ROOT',
  REPORTS_GROUP_REPORT_QUIZ_QUESTION_PAGE_ROOT: 'REPORTS_GROUP_REPORT_QUIZ_QUESTION_PAGE_ROOT',
  REPORTS_QUIZ_QUESTION_PAGE_ROOT: 'REPORTS_QUIZ_QUESTION_PAGE_ROOT',
  REPORTS_LESSON_PRACTICE_QUIZ_LEARNER_PAGE_ROOT: 'REPORTS_LESSON_PRACTICE_QUIZ_LEARNER_PAGE_ROOT',
};

export const GroupModals = {
  CREATE_GROUP: 'CREATE_GROUP',
  RENAME_GROUP: 'RENAME_GROUP',
  DELETE_GROUP: 'DELETE_GROUP',
};

export const pageNameToModuleMap = {
  [PageNames.EXAMS]: 'examsRoot',
  [PageNames.EXAM_REPORT]: 'examReport',
  [PageNames.EXAM_REPORT_DETAIL]: 'examReportDetail',
  [LessonsPageNames.PLAN_LESSONS_ROOT]: 'lessonsRoot',
  [LessonsPageNames.RESOURCE_USER_REPORT]: 'exerciseDetail',
  // Omitting modules for resource selection, exam creation, and preview to prevent
  // default module state resetting behavior.
};
