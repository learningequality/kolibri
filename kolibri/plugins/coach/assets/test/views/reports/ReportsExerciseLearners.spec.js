import { shallowMount, mount } from '@vue/test-utils';

import { STATUSES } from '../../../src/modules/classSummary/constants';
import ReportsExerciseLearners from '../../../src/views/reports/ReportsExerciseLearners';

const entries = [
  {
    id: 'd4b',
    name: 'learner1',
    username: 'learner1',
    groups: [{ id: 'dc2', name: 'group1' }, { id: '23s', name: 'group2' }],
    exerciseLearnerLink: '#/2e3/reports/lessons/79b/exercises/a97/learners/d4b',
    statusObj: {
      learner_id: 'd4b',
      content_id: 'a97',
      status: STATUSES.completed,
      last_activity: new Date('2019-05-04T07:00:00Z'),
      time_spent: 92.5,
    },
  },
  {
    id: 'a5d',
    name: 'learner2',
    username: 'learner2',
    groups: [{ id: '23s', name: 'group2' }],
    exerciseLearnerLink: '#/2e3/reports/lessons/79b/exercises/a97/learners/a5d',
    statusObj: {
      learner_id: 'a5d',
      content_id: 'a97',
      status: STATUSES.started,
      last_activity: new Date('2019-05-05T08:00:00Z'),
      time_spent: 17.14,
    },
  },
];

const initWrapper = propsData => {
  return mount(ReportsExerciseLearners, {
    propsData,
    stubs: {
      KRouterLink: {
        props: ['to', 'text'],
        template: '<div>{{ text }}</div>',
      },
    },
  });
};

const getCol = (row, colIndex) => {
  return row.findAll('td').at(colIndex);
};

jest.mock('kolibri.utils.serverClock', () => {
  return {
    now: () => new Date('2019-05-05T11:00:00Z'),
  };
});

describe('ReportsExerciseLearners', () => {
  it('smoke test', () => {
    const wrapper = shallowMount(ReportsExerciseLearners);

    expect(wrapper.isVueInstance()).toBe(true);
  });

  it('renders all entries', () => {
    const wrapper = initWrapper({
      entries,
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it("doesn't render groups information when show groups set to false", () => {
    const wrapper = initWrapper({
      entries,
      showGroupsColumn: false,
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  describe('when an exercise has started', () => {
    let row;

    beforeAll(() => {
      const wrapper = initWrapper({
        entries: [
          {
            name: 'learner1',
            groups: [],
            exerciseLearnerLink: '#/2e3/reports/lessons/79b/exercises/a97/learners/d4b',
            statusObj: {
              status: STATUSES.completed,
              last_activity: new Date('2019-05-05T08:13:22Z'),
              time_spent: 17.14,
            },
          },
        ],
      });

      row = wrapper.find('[data-test="entry"]');
    });

    it("renders learner's name as a link to an exercise", () => {
      expect(getCol(row, 0).contains({ name: 'KRouterLink' })).toBe(true);

      const link = getCol(row, 0).find({ name: 'KRouterLink' });
      expect(link.props().to).toEqual('#/2e3/reports/lessons/79b/exercises/a97/learners/d4b');
    });

    it('renders time spent', () => {
      expect(getCol(row, 2).text()).not.toBe('—');
    });

    it('renders last activity time', () => {
      expect(getCol(row, 4).text()).not.toBe('—');
    });
  });

  describe('when an exercise has not yet started', () => {
    let row;

    beforeAll(() => {
      const wrapper = initWrapper({
        entries: [
          {
            name: 'learner1',
            groups: [],
            exerciseLearnerLink: '#/2e3/reports/lessons/79b/exercises/a97/learners/d4b',
            statusObj: {
              status: STATUSES.notStarted,
            },
          },
        ],
      });

      row = wrapper.find('[data-test="entry"]');
    });

    it("doesn't render learner's as a link", () => {
      expect(getCol(row, 0).contains({ name: 'KRouterLink' })).toBe(false);
    });

    it("doesn't render time spent", () => {
      expect(getCol(row, 2).text()).toBe('—');
    });

    it("doesn't render last activity time", () => {
      expect(getCol(row, 4).text()).toBe('—');
    });
  });
});
