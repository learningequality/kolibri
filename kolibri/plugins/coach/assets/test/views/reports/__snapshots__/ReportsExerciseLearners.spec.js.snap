// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReportsExerciseLearners doesn't render groups information when show groups set to false 1`] = `
<div class="core-table-container">
  <table class="core-table">
    <thead style="border-bottom: 1px solid #dedede; font-size: 12px; color: rgb(97, 97, 97);">
      <tr>
        <th>Name</th>
        <th>Progress</th>
        <th>Time spent</th>
        <!---->
        <th>Last activity</th>
      </tr>
    </thead><span tag="tbody" name="list"><tr data-test="entry"><td><span class="labeled-icon-wrapper"><div class="icon"><mat-svg name="person" category="social" style="fill: #212121;"></mat-svg></div> <div class="label"><span dir="auto" class="debug-warning"><a dir="auto" class="KRouterLink-noKey-0_1eg6lrk link">
  learner1
</a></span>
</div></span></td>
<td>
  <div><span class="labeled-icon-wrapper" nowrap="nowrap"><div class="icon"><mat-svg name="stars" category="action" style="fill: #ffc107;"></mat-svg></div> <div class="label"><span dir="auto" class="debug-warning">
    Completed
  </span></div></span>
  <!---->
  </div>
</td>
<td><span>92 seconds</span></td>
<!---->
<td><span>
  yesterday
</span></td>
</tr>
<tr data-test="entry">
  <td><span class="labeled-icon-wrapper"><div class="icon"><mat-svg name="person" category="social" style="fill: #212121;"></mat-svg></div> <div class="label"><span dir="auto" class="debug-warning"><a dir="auto" class="KRouterLink-noKey-0_1eg6lrk link">
  learner2
</a></span></div></span></td>
  <td>
    <div><span class="labeled-icon-wrapper" nowrap="nowrap"><div class="icon"><mat-svg name="access_time" category="device" style="fill: #03a9f4;"></mat-svg></div> <div class="label"><span dir="auto" class="debug-warning">
    Started
  </span></div></span>
    <!---->
    </div>
  </td>
  <td><span>17 seconds</span></td>
  <!---->
  <td><span>
  3 hours ago
</span></td>
</tr></span></table>
</div>
`;

exports[`ReportsExerciseLearners renders all entries 1`] = `
<div class="core-table-container">
  <table class="core-table">
    <thead style="border-bottom: 1px solid #dedede; font-size: 12px; color: rgb(97, 97, 97);">
      <tr>
        <th>Name</th>
        <th>Progress</th>
        <th>Time spent</th>
        <th>
          Groups
        </th>
        <th>Last activity</th>
      </tr>
    </thead><span tag="tbody" name="list"><tr data-test="entry"><td><span class="labeled-icon-wrapper"><div class="icon"><mat-svg name="person" category="social" style="fill: #212121;"></mat-svg></div> <div class="label"><span dir="auto" class="debug-warning"><a dir="auto" class="KRouterLink-noKey-0_1eg6lrk link">
  learner1
</a></span>
</div></span></td>
<td>
  <div><span class="labeled-icon-wrapper" nowrap="nowrap"><div class="icon"><mat-svg name="stars" category="action" style="fill: #ffc107;"></mat-svg></div> <div class="label"><span dir="auto" class="debug-warning">
    Completed
  </span></div></span>
  <!---->
  </div>
</td>
<td><span>92 seconds</span></td>
<td>
  <div class="items-label"><span>
    group1, group2
  </span></div>
</td>
<td><span>
  yesterday
</span></td>
</tr>
<tr data-test="entry">
  <td><span class="labeled-icon-wrapper"><div class="icon"><mat-svg name="person" category="social" style="fill: #212121;"></mat-svg></div> <div class="label"><span dir="auto" class="debug-warning"><a dir="auto" class="KRouterLink-noKey-0_1eg6lrk link">
  learner2
</a></span></div></span></td>
  <td>
    <div><span class="labeled-icon-wrapper" nowrap="nowrap"><div class="icon"><mat-svg name="access_time" category="device" style="fill: #03a9f4;"></mat-svg></div> <div class="label"><span dir="auto" class="debug-warning">
    Started
  </span></div></span>
    <!---->
    </div>
  </td>
  <td><span>17 seconds</span></td>
  <td>
    <div class="items-label"><span>
    group2
  </span></div>
  </td>
  <td><span>
  3 hours ago
</span></td>
</tr></span></table>
</div>
`;
